src/core/net/tcp6.cpp

```diff
@@ -302,27 +456,222 @@ Error Tcp::Listener::Deinitialize(void)
     return error;
 }
 
-Error Tcp::HandleMessage(ot::Ip6::Header &aIp6Header, Message &aMessage, MessageInfo &aMessageInfo)
+bool Tcp::Listener::Matches(const MessageInfo &aMessageInfo) const
 {
-    OT_UNUSED_VARIABLE(aIp6Header);
-    OT_UNUSED_VARIABLE(aMessage);
-    OT_UNUSED_VARIABLE(aMessageInfo);
+    bool matches = false;
+
+    VerifyOrExit(mTcbListen.t_state == TCP6S_LISTEN);
+    VerifyOrExit(mTcbListen.lport == HostSwap16(aMessageInfo.GetSockPort()));
+    VerifyOrExit(GetLocalIp6Address().IsUnspecified() || GetLocalIp6Address() == aMessageInfo.GetSockAddr());
+
+    matches = true;
+
+exit:
+    return matches;
+}
 
+Error Tcp::HandleMessage(ot::Ip6::Header &aIp6Header, Message &aMessage, MessageInfo &aMessageInfo)
+{
     Error error = kErrorNotImplemented;
 
-    for (Endpoint &active : mEndpoints)
+    /*
+     * The type uint32_t was chosen for alignment purposes. The size is the
+     * maximum TCP header size, including options.
+     */
+    uint32_t header[15];
+
+    uint16_t length = aIp6Header.GetPayloadLength();
+    uint8_t  headerSize;
+
+    struct ip6_hdr *ip6Header;
+    struct tcphdr * tcpHeader;
+
+    Endpoint *endpoint;
+    Endpoint *endpointPrev;
+
+    Listener *listener;
+    Listener *listenerPrev;
+
+    VerifyOrExit(length == aMessage.GetLength() - aMessage.GetOffset(), error = kErrorParse);
+    VerifyOrExit(length >= sizeof(Tcp::Header), error = kErrorParse);
+    SuccessOrExit(error = aMessage.Read(aMessage.GetOffset() + offsetof(struct tcphdr, th_off_x2), &headerSize, 1));
+    headerSize = static_cast<uint8_t>((headerSize >> TH_OFF_SHIFT) << 2);
+    VerifyOrExit(headerSize >= sizeof(struct tcphdr) && headerSize <= sizeof(header) &&
+                     static_cast<uint16_t>(headerSize) <= length,
+                 error = kErrorParse);
+    SuccessOrExit(error = Checksum::VerifyMessageChecksum(aMessage, aMessageInfo, kProtoTcp));
+    SuccessOrExit(error = aMessage.Read(aMessage.GetOffset(), &header[0], headerSize));
+
+    ip6Header = reinterpret_cast<struct ip6_hdr *>(&aIp6Header);
+    tcpHeader = reinterpret_cast<struct tcphdr *>(&header[0]);
+    tcp_fields_to_host(tcpHeader);
+
+    aMessageInfo.mPeerPort = HostSwap16(tcpHeader->th_sport);
+    aMessageInfo.mSockPort = HostSwap16(tcpHeader->th_dport);
+
+    endpoint = mEndpoints.FindMatching(aMessageInfo, endpointPrev);
+    if (endpoint != nullptr)
     {
-        OT_UNUSED_VARIABLE(active);
+        struct signals  sig;
+        otLinkedBuffer *priorHead = lbuf_head(&endpoint->mTcb.sendbuf);
+        memset(&sig, 0x00, sizeof(sig));
+        int nextAction = tcp_input(ip6Header, tcpHeader, &aMessage, &endpoint->mTcb, nullptr, &sig);
```

Comment:
Move with other vairable at the top of scope?

Reply:
OK. Fixed in the latest update.